### ISSUE TITLE
Emformer forward and infer output difference

### ISSUE THREAD

**Issue by Tomiinek** [pytorch/audio#2200]
### 🐛 Describe the bug

Hello (@hwangjeff ), I noticed discrepancy between Emformer `forward` and `infer` outputs when using non-zero right context:

``` python

encoder_right_context_size = 1
encoder_left_context_size = 4
chunk_size = 4

emformer = Emformer(
    input_dim = 8,
    num_heads = 1,
    ffn_dim = 8,
    num_layers = 1,
    left_context_length = encoder_left_context_size,
    right_context_length = encoder_right_context_size,
    segment_length = chunk_size
)
emformer = emformer.eval()

x = torch.randn(1, 100, 8) 
x_lengths = torch.tensor([100]).long() # something realyyyyyy long so that we do not have to bother with paddings or lengths

offline_output, _ = emformer(x, x_lengths)

online_states = None
for i in range(0, x.shape[1], chunk_size):
    s = i
    e = i + chunk_size 
    x_chunk = x[:, s:e]
    
    online_output, _, online_states = emformer.infer(x_chunk, torch.ones(x_chunk.shape[0], device=x.device) * x_chunk.shape[1], online_states)
    break # we do not need to continue, but we could :) 
    
print(offline_output[:, :chunk_size - encoder_right_context_size] - online_output)
```

Ouputs something like: 
```
tensor([[
        [ 0.0156, -0.1142,  0.0152, -0.1302, -0.0019,  0.0077,  0.1108, 0.0971],
        [ 0.0376,  0.0665, -0.0177, -0.2270, -0.0756, -0.1156,  0.1082, 0.2236],
        [-0.0721,  0.0115,  0.0060, -0.2362,  0.1189, -0.0636,  0.0615,  0.1741]
]], grad_fn=<SubBackward0>)
```

I would expect having `offline_output` and `online_output` exactly the same values which is true only if `encoder_right_context_size = 0`. What am I missing or doing wrong? The difference is substantial on real data.

### Versions

The current version in `main`

**Comment by hwangjeff** [comment]
@Tomiinek thanks for flagging! There are a few things to point out:
- `segment_length` is exclusive of `right_context_length`.
- Accordingly, the input to `infer` should contain the right context and be of length `segment_length + right_context_length`.
- Also, for `forward`, the `length` input should exclude the length of the right context, whereas for `infer` it should include it. This is admittedly a little confusing, and I will update the documentation to clarify this.

Accounting for these in edits to the code you provided, we have
```
import torch
from torchaudio.models import Emformer

encoder_right_context_size = 1
encoder_left_context_size = 4
chunk_size = 4

emformer = Emformer(
    input_dim=8,
    num_heads=1,
    ffn_dim=8,
    num_layers=1,
    left_context_length=encoder_left_context_size,
    right_context_length=encoder_right_context_size,
    segment_length=chunk_size,
)
emformer = emformer.eval()

x = torch.randn(1, 100, 8)
x_lengths = torch.tensor([100 - encoder_right_context_size]).long()

offline_output, _ = emformer(x, x_lengths)

online_states = None
for i in range(0, x.shape[1], chunk_size):
    s = i
    e = i + chunk_size + encoder_right_context_size
    x_chunk_rc = x[:, s:e]

    online_output, _, online_states = emformer.infer(
        x_chunk_rc, torch.ones(x_chunk_rc.shape[0], device=x.device) * x_chunk_rc.shape[1], online_states
    )
    break

print(offline_output[:, : chunk_size] - online_output)
```

, which yields a much more reasonable output:
```
tensor([[[ 1.1921e-07,  2.9802e-08, -2.9802e-08,  5.9605e-08,  1.1921e-07,
           2.9802e-08,  1.1921e-07,  1.1921e-07],
         [ 1.1921e-07, -5.9605e-08,  0.0000e+00,  0.0000e+00, -1.1921e-07,
           0.0000e+00,  0.0000e+00,  0.0000e+00],
         [ 0.0000e+00,  0.0000e+00,  0.0000e+00,  6.7055e-08,  3.5763e-07,
          -1.1921e-07,  0.0000e+00,  5.9605e-08],
         [ 1.1921e-07,  5.9605e-08,  4.4703e-08, -1.1921e-07,  5.9605e-08,
           5.9605e-08,  1.1921e-07,  1.0431e-07]]], grad_fn=<SubBackward0>)
```

Let us know if you any other questions or thoughts.

**Comment by Tomiinek** [comment]
Thanks for the prompt reply! Makes sense, my fault :slightly_smiling_face:

**Comment by hwangjeff** [comment]
No problem!

**Comment by Tomiinek** [comment]
If you don't mind I will just describe my reasoning, maybe it help somebody who is in a late-night debugging rage as I was, or could help to clarify the documentation :slightly_smiling_face: 

In `Emformer` constructor, there are:
```
right_context_length (int, optional) – length of right context. (Default: 0)
segment_length (int, optional) – length of each input segment. (Default: 128)
```
In the `infer` method, there are:
```
input ([torch.Tensor]) – utterance frames right-padded with right context frames, with shape (B, T, D).
```
And returns:
```
output frames, with shape (B, T - ``right_context_length`, D)`.
```

From this, I deduced that `segment_length` corresponds to `T` because of the mention "length of each **input** segment".
However, it should correspond to `T - right_context_length` exactly.

Maybe some asserts in `infer` could be helpful and made the code bullet (or my stupidity :smile:) proof? If I am not wrong, currently, it seems that it is possible to push through it arbitrarily-sized chunks.

**Comment by hwangjeff** [comment]
@Tomiinek this is very helpful — thanks! Sorry to hear about the debugging issues; I see where the confusion lies and agree that adjusting the documentation and adding some asserts would make this more usable. Will make the changes.

**Comment by hwangjeff** [comment]
Addressed in #2207
